core/include/scipp/core/dtype.h

```diff
@@ -57,6 +58,76 @@ constexpr DType dtype<sparse_container<int32_t>> = DType::SparseInt32;
+template <> struct TypeMap<DType::Bool> { using type = Bool; };
+template <> struct TypeMap<DType::SparseDouble> {
+  using type = sparse_container<double>;
+};
+template <> struct TypeMap<DType::SparseFloat> {
+  using type = sparse_container<float>;
+};
+template <> struct TypeMap<DType::SparseInt64> {
+  using type = sparse_container<int64_t>;
+};
+template <> struct TypeMap<DType::SparseInt32> {
+  using type = sparse_container<int32_t>;
+};
+template <> struct TypeMap<DType::Dataset> { using type = Dataset; };
+template <> struct TypeMap<DType::EigenVector3d> {
+  using type = Eigen::Vector3d;
+};
+
+template <class T1, class T2>
+constexpr bool Enable =
+    std::is_same_v<T1, T2> || std::is_trivially_constructible_v<T1, T2>;
+
+template <class T1, class T2, bool = Enable<T1, T2>> struct castHelper {};
+
+template <class T1, class T2> struct castHelper<T1, T2, false> {
+  static T1 cast(const T2 &arg) {
+    throw std::runtime_error(std::string("Can't cast: ") + typeid(arg).name() +
+                             " and " + typeid(T1{}).name());
+  }
+};
+
+template <class T1, class T2> struct castHelper<T1, T2, true> {
+  static T1 cast(const T2 &arg) { return T1(arg); }
+};
+
+#define MAKE_VARIABLE_CASE(tp, x)                                              \
+  case DType::tp:                                                              \
+    return makeVariable<TypeMap<DType::tp>::type>(                             \
+        castHelper<TypeMap<DType::tp>::type, decltype(x)>::cast(x));           \
+    break;
+
+#define MAKE_VARIABLE_DTYPED(tp, x)                                            \
+  switch (tp) {                                                                \
+    MAKE_VARIABLE_CASE(Double, x)                                              \
+    MAKE_VARIABLE_CASE(Float, x)                                               \
+    MAKE_VARIABLE_CASE(Int32, x)                                               \
+    MAKE_VARIABLE_CASE(Int64, x)                                               \
+    MAKE_VARIABLE_CASE(String, x)                                              \
+    MAKE_VARIABLE_CASE(Bool, x)                                                \
+    MAKE_VARIABLE_CASE(SparseDouble, x)                                        \
+    MAKE_VARIABLE_CASE(SparseFloat, x)                                         \
+    MAKE_VARIABLE_CASE(SparseInt64, x)                                         \
+    MAKE_VARIABLE_CASE(SparseInt32, x)                                         \
+    MAKE_VARIABLE_CASE(Dataset, x)                                             \
+    MAKE_VARIABLE_CASE(EigenVector3d, x)                                       \
+  case DType::Unknown:                                                         \
+    throw std::logic_error("Can't convert to Unknown type");                   \
+    return {};                                                                 \
```

Comment:
I think the macros and listing all cases can be avoided using the `CallDType` template, see, e.g., https://github.com/scipp/scipp/blob/10543377cbfcd9b17e873cabdd50d747ab3322a4/python/variable.cpp#L116-L117.

When you use that, you can probably also remove a lot of the added helper code in `dtype.h`?

We can consider improving `CallDType` so there is an way to call something for "all known types", without listing them at every call site.

Reply:
The idea of using macro is not to write the special `CallDType` for different types of functions. But now the implementation covers only `makeVariable` case. We probably should think about this, but let extract the separate issue.

Reply:
The *macro* is specific to `makeVariable`. `CallDType` accepts a functor. You can avoid a lot of code if you use it, instead of these lengthy macros.